File: sdks/javascript/sdk/test/deviceId.test.ts
import { describe, it, expect, beforeEach, afterEach } from "vitest";
import { existsSync, mkdtempSync, readFileSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, isAbsolute } from "node:path";
import {
  telemetryDisabled,
  resolveDeviceId,
  resolveClientDeviceId,
  defaultDeviceIdDir,
  applyDeviceId,
  applyDeviceIdOnce,
  type DeviceIdState,
} from "../src/utils/deviceId";

const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

describe("deviceId", () => {
  let dir: string;
  beforeEach(() => {
    dir = mkdtempSync(join(tmpdir(), "moss-devid-"));
  });
  afterEach(() => {
    rmSync(dir, { recursive: true, force: true });
  });

  // --- resolveDeviceId ---

  it("creates and persists a UUID on first resolve", () => {
    const id = resolveDeviceId(dir, {});
    expect(id).toMatch(UUID_RE);
    const file = join(dir, ".moss-device-id");
    expect(existsSync(file)).toBe(true);
    expect(readFileSync(file, "utf8").trim()).toBe(id);
  });

  it("returns the same id on subsequent resolves", () => {
    const first = resolveDeviceId(dir, {});
    const second = resolveDeviceId(dir, {});
    expect(second).toBe(first);
  });

  it("honors a pre-seeded id file", () => {
    writeFileSync(join(dir, ".moss-device-id"), "preseeded-id-1", "utf8");
    expect(resolveDeviceId(dir, {})).toBe("preseeded-id-1");
  });

  it("returns undefined when telemetry is disabled", () => {
    expect(resolveDeviceId(dir, { MOSS_DISABLE_TELEMETRY: "true" })).toBeUndefined();
    expect(existsSync(join(dir, ".moss-device-id"))).toBe(false);
  });

  it("telemetryDisabled parses common truthy values", () => {
    expect(telemetryDisabled({ MOSS_DISABLE_TELEMETRY: "1" })).toBe(true);
    expect(telemetryDisabled({ MOSS_DISABLE_TELEMETRY: "TRUE" })).toBe(true);
    expect(telemetryDisabled({})).toBe(false);
    expect(telemetryDisabled({ MOSS_DISABLE_TELEMETRY: "0" })).toBe(false);
  });

  // --- defaultDeviceIdDir ---

  it("defaultDeviceIdDir resolves to <home>/.moss from $HOME", () => {
    expect(defaultDeviceIdDir({ HOME: "/home/u" })).toBe(join("/home/u", ".moss"));
  });

  it("defaultDeviceIdDir falls back to %USERPROFILE% (Windows)", () => {
    expect(defaultDeviceIdDir({ USERPROFILE: "C:\\Users\\u" })).toBe(
      join("C:\\Users\\u", ".moss"),
    );
  });

  it("defaultDeviceIdDir ignores a blank $HOME (would otherwise resolve to CWD)", () => {
    // A blank HOME must fall through, not produce a relative ".moss".
    expect(defaultDeviceIdDir({ HOME: "  ", USERPROFILE: "C:\\Users\\u" })).toBe(
      join("C:\\Users\\u", ".moss"),
    );
    expect(isAbsolute(defaultDeviceIdDir({ HOME: "" }))).toBe(true);
  });

  // --- resolveClientDeviceId ---

  it("resolveClientDeviceId persists under cachePath when provided", () => {
    const state: DeviceIdState = { applied: false };
    const id = resolveClientDeviceId(state, dir, {});
    expect(id).toMatch(UUID_RE);
    expect(readFileSync(join(dir, ".moss-device-id"), "utf8").trim()).toBe(id);
    expect(state.id).toBe(id);
  });

  it("resolveClientDeviceId falls back to <home>/.moss when no cachePath", () => {
    const state: DeviceIdState = { applied: false };
    const id = resolveClientDeviceId(state, undefined, { HOME: dir });
    expect(id).toMatch(UUID_RE);
    expect(readFileSync(join(dir, ".moss", ".moss-device-id"), "utf8").trim()).toBe(id);
  });

  it("resolveClientDeviceId memoizes the first id across calls and locations", () => {
    const state: DeviceIdState = { applied: false };
    const first = resolveClientDeviceId(state, dir, {});
    // A later call with no cachePath must reuse the memoized id, not mint a
    // second one from the fallback dir — one device, one id.
    const second = resolveClientDeviceId(state, undefined, { HOME: join(dir, "other") });
    expect(second).toBe(first);
    expect(existsSync(join(dir, "other", ".moss", ".moss-device-id"))).toBe(false);
  });

  it("resolveClientDeviceId returns undefined and does not memoize when telemetry disabled", () => {
    const state: DeviceIdState = { applied: false };
    expect(resolveClientDeviceId(state, dir, { MOSS_DISABLE_TELEMETRY: "1" })).toBeUndefined();
    expect(state.id).toBeUndefined();
  });

  it("resolveClientDeviceId honors a runtime telemetry-disable even after an id was memoized", () => {
    const state: DeviceIdState = { id: "preset-id", applied: true };
    expect(resolveClientDeviceId(state, dir, { MOSS_DISABLE_TELEMETRY: "1" })).toBeUndefined();
  });

  it("resolveClientDeviceId treats a blank cachePath as absent and uses the fallback dir", () => {
    const state: DeviceIdState = { applied: false };
    const id = resolveClientDeviceId(state, "   ", { HOME: dir });
    expect(id).toMatch(UUID_RE);
    // Persisted to the fallback dir, not the CWD (`resolve("   ")`).
    expect(readFileSync(join(dir, ".moss", ".moss-device-id"), "utf8").trim()).toBe(id);
  });

  // --- applyDeviceId ---

  it("applyDeviceId pushes the id to the target and reports success", () => {
    const calls: string[] = [];
    const ok = applyDeviceId({ setDeviceId: (d: string) => calls.push(d) }, "abc");
    expect(ok).toBe(true);
    expect(calls).toEqual(["abc"]);
  });

  it("applyDeviceId treats a missing setDeviceId (older binding) as terminal success", () => {
    const target = {} as unknown as { setDeviceId: (d: string) => void };
    expect(applyDeviceId(target, "abc")).toBe(true);
  });

  it("applyDeviceId reports failure when setDeviceId throws (so the caller can retry)", () => {
    const target = {
      setDeviceId: () => {
        throw new Error("transient binding error");
      },
    };
    expect(applyDeviceId(target, "abc")).toBe(false);
  });

  // --- applyDeviceIdOnce ---

  it("applyDeviceIdOnce sets the id exactly once and memoizes", () => {
    const calls: string[] = [];
    const target = { setDeviceId: (d: string) => calls.push(d) };
    const state: DeviceIdState = { applied: false };

    applyDeviceIdOnce(target, state, dir, {});
    applyDeviceIdOnce(target, state, dir, {});
    expect(calls.length).toBe(1);
    expect(calls[0]).toMatch(UUID_RE);
    expect(state.id).toBe(calls[0]);
  });

  it("applyDeviceIdOnce falls back to the default device-id dir when no cachePath", () => {
    const calls: string[] = [];
    const target = { setDeviceId: (d: string) => calls.push(d) };
    const state: DeviceIdState = { applied: false };
    applyDeviceIdOnce(target, state, undefined, { HOME: dir });
    expect(calls.length).toBe(1);
    expect(calls[0]).toMatch(UUID_RE);
    expect(state.applied).toBe(true);
    expect(existsSync(join(dir, ".moss", ".moss-device-id"))).toBe(true);
  });

  it("applyDeviceIdOnce does nothing when telemetry disabled", () => {
    const calls: string[] = [];
    const target = { setDeviceId: (d: string) => calls.push(d) };
    const state: DeviceIdState = { applied: false };
    applyDeviceIdOnce(target, state, dir, { MOSS_DISABLE_TELEMETRY: "yes" });
    expect(calls.length).toBe(0);
  });

  it("applyDeviceIdOnce marks applied without throwing when setDeviceId is absent (older binding)", () => {
    const target = {} as unknown as { setDeviceId: (d: string) => void };
    const state: DeviceIdState = { applied: false };
    expect(() => applyDeviceIdOnce(target, state, dir, {})).not.toThrow();
    expect(state.applied).toBe(true);
  });

  it("applyDeviceIdOnce leaves applied=false (retries) when setDeviceId throws", () => {
    let calls = 0;
    const target = {
      setDeviceId: () => {
        calls++;
        throw new Error("transient binding error");
      },
    };
    const state: DeviceIdState = { applied: false };
    expect(() => applyDeviceIdOnce(target, state, dir, {})).not.toThrow();
    expect(state.applied).toBe(false);
    applyDeviceIdOnce(target, state, dir, {}); // retried on next call
    expect(calls).toBe(2);
  });
});

File: sdks/elixir/sdk/test/moss/device_id_test.exs
defmodule Moss.DeviceIdTest do
  @moduledoc """
  SDK-layer tests for the MOS-14 device-id contract. These exercise
  sourcing / persistence / memoization / opt-out / best-effort apply against a
  fake apply function — exactly as the TS reference `deviceId.test.ts` uses a
  fake `setDeviceId` target — so the contract is verifiable independently of
  the (not-yet-existing) `set_device_id` NIF.
  """
  use ExUnit.Case, async: true

  alias Moss.DeviceId

  @uuid_re ~r/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i

  setup do
    dir = Path.join(System.tmp_dir!(), "moss-devid-#{System.unique_integer([:positive])}")
    File.mkdir_p!(dir)
    on_exit(fn -> File.rm_rf(dir) end)
    {:ok, dir: dir}
  end

  # --- resolve/2 (R1, R2.2, R2.4) ---

  test "creates and persists a UUID on first resolve", %{dir: dir} do
    id = DeviceId.resolve(dir, %{})
    assert id =~ @uuid_re
    file = Path.join(dir, ".moss-device-id")
    assert File.exists?(file)
    assert File.read!(file) |> String.trim() == id
  end

  test "returns the same id on subsequent resolves", %{dir: dir} do
    first = DeviceId.resolve(dir, %{})
    assert DeviceId.resolve(dir, %{}) == first
  end

  test "honors a pre-seeded id file", %{dir: dir} do
    File.write!(Path.join(dir, ".moss-device-id"), "preseeded-id-1")
    assert DeviceId.resolve(dir, %{}) == "preseeded-id-1"
  end

  test "treats a blank persisted value as absent and regenerates", %{dir: dir} do
    File.write!(Path.join(dir, ".moss-device-id"), "   \n")
    id = DeviceId.resolve(dir, %{})
    assert id =~ @uuid_re
  end

  test "returns nil when telemetry is disabled and writes nothing", %{dir: dir} do
    assert DeviceId.resolve(dir, %{"MOSS_DISABLE_TELEMETRY" => "true"}) == nil
    refute File.exists?(Path.join(dir, ".moss-device-id"))
  end

  # --- telemetry_disabled?/1 (R4.1) ---

  test "telemetry_disabled? parses common truthy values" do
    assert DeviceId.telemetry_disabled?(%{"MOSS_DISABLE_TELEMETRY" => "1"})
    assert DeviceId.telemetry_disabled?(%{"MOSS_DISABLE_TELEMETRY" => "TRUE"})
    assert DeviceId.telemetry_disabled?(%{"MOSS_DISABLE_TELEMETRY" => " Yes "})
    refute DeviceId.telemetry_disabled?(%{})
    refute DeviceId.telemetry_disabled?(%{"MOSS_DISABLE_TELEMETRY" => "0"})
  end

  # --- resolve_client/3 (R3, R2.3) ---

  test "resolve_client persists under cache_path when provided", %{dir: dir} do
    {id, state} = DeviceId.resolve_client(DeviceId.new_state(), dir, %{})
    assert id =~ @uuid_re
    assert File.read!(Path.join(dir, ".moss-device-id")) |> String.trim() == id
    assert state.id == id
  end

  test "resolve_client memoizes the first id across calls and locations", %{dir: dir} do
    {first, state} = DeviceId.resolve_client(DeviceId.new_state(), dir, %{})
    # A later call with no cache_path must reuse the memoized id — one device, one id.
    {second, _state} =
      DeviceId.resolve_client(state, nil, %{"HOME" => Path.join(dir, "other")})

    assert second == first
    refute File.exists?(Path.join([dir, "other", ".moss", ".moss-device-id"]))
  end

  test "resolve_client returns nil and does not memoize when telemetry disabled", %{dir: dir} do
    {id, state} =
      DeviceId.resolve_client(DeviceId.new_state(), dir, %{"MOSS_DISABLE_TELEMETRY" => "1"})

    assert id == nil
    assert state.id == nil
  end

  test "resolve_client honors runtime disable even after an id was memoized", %{dir: dir} do
    state = %{id: "preset-id", applied: true}
    {id, _} = DeviceId.resolve_client(state, dir, %{"MOSS_DISABLE_TELEMETRY" => "1"})
    assert id == nil
  end

  test "resolve_client treats a blank cache_path as absent, uses fallback dir", %{dir: dir} do
    {id, _} = DeviceId.resolve_client(DeviceId.new_state(), "   ", %{"HOME" => dir})
    assert id =~ @uuid_re
    # persisted under the fallback dir, not the CWD
    assert File.read!(Path.join([dir, ".moss", ".moss-device-id"])) |> String.trim() == id
  end

  # --- apply/2 (R5.3, R5.4) ---

  test "apply pushes the id to the target and reports success" do
    parent = self()
    ok? = DeviceId.apply(fn id -> send(parent, {:applied, id}) && :ok end, "abc")
    assert ok?
    assert_received {:applied, "abc"}
  end

  test "apply treats a nil apply_fun (older/absent binding) as terminal success" do
    assert DeviceId.apply(nil, "abc")
  end

  test "apply treats :unsupported as terminal success" do
    assert DeviceId.apply(fn _ -> :unsupported end, "abc")
  end

  test "apply reports failure when apply_fun raises (so caller can retry)" do
    refute DeviceId.apply(fn _ -> raise "transient binding error" end, "abc")
  end

  test "apply reports failure on {:error, _}" do
    refute DeviceId.apply(fn _ -> {:error, :nope} end, "abc")
  end

  # --- apply_once/4 (R3.2, R3.3) ---

  test "apply_once sets the id exactly once and memoizes", %{dir: dir} do
    parent = self()
    fun = fn id -> send(parent, {:call, id}) && :ok end

    state = DeviceId.apply_once(DeviceId.new_state(), fun, dir, %{})
    state = DeviceId.apply_once(state, fun, dir, %{})

    assert state.applied
    assert_received {:call, id}
    assert id =~ @uuid_re
    refute_received {:call, _}
  end

  test "apply_once does nothing when telemetry disabled", %{dir: dir} do
    parent = self()
    fun = fn id -> send(parent, {:call, id}) && :ok end
    state = DeviceId.apply_once(DeviceId.new_state(), fun, dir, %{"MOSS_DISABLE_TELEMETRY" => "yes"})
    refute state.applied
    refute_received {:call, _}
  end

  test "apply_once marks applied without raising when apply_fun is nil", %{dir: dir} do
    state = DeviceId.apply_once(DeviceId.new_state(), nil, dir, %{})
    assert state.applied
  end

  test "apply_once leaves applied=false (retries) when apply_fun raises", %{dir: dir} do
    parent = self()
    fun = fn _ -> send(parent, :called) && raise("transient") end

    state = DeviceId.apply_once(DeviceId.new_state(), fun, dir, %{})
    refute state.applied
    _ = DeviceId.apply_once(state, fun, dir, %{})
    # called twice: retried on the second call
    assert_received :called
    assert_received :called
  end
end
